fsm.go

```diff
@@ -306,13 +326,17 @@ func (t transitionerStruct) transition(f *FSM) error {
 // general version.
 func (f *FSM) beforeEventCallbacks(e *Event) error {
 	if fn, ok := f.callbacks[cKey{e.Event, callbackBeforeEvent}]; ok {
+		f.mutex.Unlock()
```

Comment:
It would be less repetitive and error prone in these cases to wrap the callbacks in `NewFSM` at line 193 with a function that does the unlocking and locking:
```
f.callbacks[cKey{target, callbackType}] = func(e *Event) {
    // This always happens in a locked condition, but needs to be unlocked
    // to not risk a deadlock. The reversed order is not a mistake!
    f.mutex.Unlock()
    defer f.mutex.Lock()
    c(e)
}
```
If you do this, could you also change `var f FSM` in the beginning to pointer form:
```
f := &FSM{
	transitionerObj: &transitionerStruct{},
	current: initial,
	transitions: make(map[eKey]string),
	callbacks: make(map[cKey]Callback),
}
```
And return it with out dereferencing.

Do you think this would be good?